src/path.cr

```diff
@@ -568,10 +570,20 @@ struct Path
 
     name = @name
 
-    if name == "~"
-      name = (home || Path.home).to_kind(@kind).normalize.to_s
-    elsif name.starts_with?("~/")
-      name = (home || Path.home).to_kind(@kind).normalize.join(name.byte_slice(2, name.bytesize - 2)).to_s
+    if home
+      home = case home
+             when Path   then home
+             when String then Path[home]
+             when true   then Path.home
+             when false  then nil
+             else             raise ArgumentError.new("Expected Path | String | Bool, not #{home.class}")
```

Comment:
This branch should be unreachable because of the argument's type restriction.

Reply:
I botched the merge slightly.  See current code where (home || Path.home) is still in use.  I have a use case where a program may use `path.expand(home: true)` and not want `Path.home` resolved.

```
download_dir = settings_from_user_editable_config_file["downloads"] || "~/Downloads"
Path[download_dir].expand(home: true)
```

"downloads" could be "/Volumes/some_drive" or /Users/Shared/Downloads".  There are many similar use cases and I think some of them quite common for intended use such as user editable configuration settings, command line flags or possibly environment variables.  I recognize the shell may expand `~/` by default but that's not where the difference is likely to be.

Example where user supplied Path expansion used by default but unlikely from a command line argument.
```
foo_path = "~/.foo"
# Path.home shouldn't run based on #8128
option_parser.on("--foo-path=[PATH]) { |arg| foo_path = arg }
Path[foo_path].expand(home: true)
```


Dropping `(home || Path.home)` within the if block would bring back #8128

Reply:
True, I forgot about the lazy evaluation. I suppose it would probably be worth to extract `home` resolution in a helper method and use that in the name branches.
```cr
private def resolve_home(home)
  case home
    when String then home = Path[home]
    when true   then home = Path.home
  end
  home.to_kind(@kind).normalize
end
```
This reads clearer than the current implementation, it centralizes the handling and removes code duplication.